File: oy_cli/agent.py
from __future__ import annotations

import time
from typing import Any

import tiktoken
import toons
from . import runtime as rt
from .providers import (
    AuthenticationError,
    AssistantMessage,
    BadRequestError,
    ChatMessage,
    PermissionDeniedError,
    RateLimitError,
    SystemMessage,
    ToolMessage,
    UserMessage,
    _tool_output_text,
)
from .runtime import active_tool_registry, session_text
from .tools import _format_todos, _positive_int, invoke_tool, tool_specs


type AgentState = dict[str, Any]
type Transcript = dict[str, Any]
type Wait = dict[str, Any]


def agent_state(
    *,
    root: rt.Path,
    tool_registry: dict[str, dict[str, Any]],
    unattended_limit_seconds: int,
    unattended_deadline: float,
    interactive: bool = False,
    approve_all_mutating_tools: bool = False,
    yolo: bool = False,
    todos: list[dict[str, str]] | None = None,
) -> AgentState:
    return {
        "root": root,
        "tool_registry": tool_registry,
        "unattended_limit_seconds": unattended_limit_seconds,
        "unattended_deadline": unattended_deadline,
        "interactive": interactive,
        "approve_all_mutating_tools": approve_all_mutating_tools,
        "yolo": yolo,
        "todos": list(todos or []),
    }


def new_agent_state(
    *,
    root: rt.Path,
    tool_registry: dict[str, dict[str, Any]],
    unattended_limit_seconds: int,
    interactive: bool = False,
    yolo: bool = False,
) -> AgentState:
    return agent_state(
        root=root,
        tool_registry=tool_registry,
        unattended_limit_seconds=unattended_limit_seconds,
        unattended_deadline=time.monotonic() + unattended_limit_seconds,
        interactive=interactive,
        yolo=yolo,
        approve_all_mutating_tools=yolo,
    )


def remaining_unattended_seconds(state: AgentState) -> float:
    return state["unattended_deadline"] - time.monotonic()


def note_progress(state: AgentState) -> None:
    if remaining_unattended_seconds(state) <= 0:
        raise TimeoutError(
            "reached unattended timeout "
            f"({rt._format_duration(state['unattended_limit_seconds'])}) without a final response"
        )


def _message_text(message: ChatMessage) -> str:
    if message.get("role") == "tool":
        return _tool_output_text(message["content"])
    return message["content"]


def count_tokens(text: str) -> int:
    return rt.count_tokens(text)


def _message_tokens(message: ChatMessage) -> int:
    return 4 + count_tokens(_message_text(message))


def _truncate_message(message: ChatMessage, max_tokens: int) -> ChatMessage:
    if message.get("role") == "tool" or not message["content"]:
        return message
    if (
        truncated := rt.truncate_str_to_tokens(
            message["content"], max_tokens=max_tokens
        )
    ) is message["content"]:
        return message
    role = message.get("role")
    if role == "system":
        return SystemMessage(truncated)
    if role == "user":
        return UserMessage(truncated)
    if role == "assistant":
        return AssistantMessage(
            truncated,
            tool_calls=message["tool_calls"],
            thought_signatures=message["thought_signatures"],
        )
    return message


def _message_role(message: ChatMessage) -> str:
    return message.get("role", "tool")


def _can_pack_message_with_toons(message: ChatMessage) -> bool:
    if message.get("role") == "user":
        return True
    if message.get("role") == "assistant":
        return not message["tool_calls"] and not message["thought_signatures"]
    return False


def _packed_history_note(messages: list[ChatMessage]) -> SystemMessage:
    packed = toons.dumps(
        {
            "messages": [
                {"role": _message_role(message), "content": _message_text(message)}
                for message in messages
            ]
        }
    )
    return SystemMessage(
        session_text("transcript", "packed_history_note", packed=packed).strip()
    )


def _pack_messages_with_toons(messages: list[ChatMessage]) -> list[ChatMessage]:
    systems = [message for message in messages if message.get("role") == "system"]
    other = [message for message in messages if message.get("role") != "system"]
    max_prefix = len(other) - 2
    if max_prefix < 2:
        return messages

    packable_prefix: list[ChatMessage] = []
    for message in other[:max_prefix]:
        if not _can_pack_message_with_toons(message):
            break
        packable_prefix.append(message)
    if len(packable_prefix) < 2:
        return messages

    best_count = 0
    best_note: SystemMessage | None = None
    raw_tokens = 0
    try:
        for count, message in enumerate(packable_prefix, start=1):
            raw_tokens += _message_tokens(message)
            if count < 2:
                continue
            note = _packed_history_note(packable_prefix[:count])
            savings = raw_tokens - _message_tokens(note)
            if savings > 0:
                best_count = count
                best_note = note
    except Exception:
        return messages

    if best_note is None:
        return messages
    return [*systems, best_note, *other[best_count:]]


def _history_units(messages: list[ChatMessage]) -> list[list[ChatMessage]]:
    units: list[list[ChatMessage]] = []
    index = 0
    while index < len(messages):
        message = messages[index]
        if message.get("role") == "assistant" and message["tool_calls"]:
            end = index + 1
            while end < len(messages) and messages[end].get("role") == "tool":
                end += 1
            units.append(messages[index:end])
            index = end
            continue
        units.append([message])
        index += 1
    return units


def transcript(
    messages: list[ChatMessage] | None = None,
    *,
    max_context_tokens: int = rt.MAX_CONTEXT_TOKENS,
    max_message_tokens: int = rt.BUDGETS["message_tokens"],
) -> Transcript:
    return {
        "messages": list(messages or []),
        "max_context_tokens": max_context_tokens,
        "max_message_tokens": max_message_tokens,
    }


def transcript_with_system_prompt(system_prompt: str) -> Transcript:
    tx = transcript()
    set_system_prompt(tx, system_prompt)
    return tx


def set_system_prompt(tx: Transcript, system_prompt: str) -> None:
    if tx["messages"] and tx["messages"][0].get("role") == "system":
        tx["messages"][0] = SystemMessage(system_prompt)
    else:
        tx["messages"][:0] = [SystemMessage(system_prompt)]


def clear_transcript(tx: Transcript, system_prompt: str) -> None:
    tx["messages"].clear()
    set_system_prompt(tx, system_prompt)


def checkpoint(tx: Transcript) -> int:
    return len(tx["messages"])


def rollback(tx: Transcript, point: int) -> None:
    del tx["messages"][point:]


def undo_last_turn(tx: Transcript) -> bool:
    for index in range(len(tx["messages"]) - 1, 0, -1):
        if tx["messages"][index].get("role") == "user":
            del tx["messages"][index:]
            return True
    return False


def add_user(tx: Transcript, prompt: str) -> None:
    tx["messages"].append(UserMessage(prompt))


def add_assistant(tx: Transcript, message: AssistantMessage) -> None:
    tx["messages"].append(message)


def add_tool_results(tx: Transcript, results: list[dict[str, Any]]) -> None:
    tx["messages"].extend(
        ToolMessage(
            tool_call_id=result["call_id"],
            name=result["name"],
            content=result["result"],
        )
        for result in results
    )


def prepared_messages(
    tx: Transcript,
    model: str | None = None,
    todos: list[dict[str, str]] | None = None,
) -> list[ChatMessage]:
    messages = [
        _truncate_message(message, tx["max_message_tokens"])
        for message in tx["messages"]
    ]
    if model:
        messages = _pack_messages_with_toons(messages)
    system_messages = [
        message for message in messages if message.get("role") == "system"
    ]
    if todos:
        system_messages.append(
            SystemMessage(
                session_text(
                    "transcript", "todo_system", todos=_format_todos(todos)
                ).strip()
            )
        )
    other = [message for message in messages if message.get("role") != "system"]
    budget = tx["max_context_tokens"] - sum(map(_message_tokens, system_messages))
    if budget <= 0:
        return system_messages
    kept_units: list[list[ChatMessage]] = []
    used = 0
    for unit in reversed(_history_units(other)):
        cost = sum(_message_tokens(message) for message in unit)
        if cost + used <= budget:
            kept_units.append(unit)
            used += cost
    kept = [message for unit in reversed(kept_units) for message in unit]
    omitted_messages = len(other) - len(kept)
    return (
        system_messages
        + (
            [
                UserMessage(
                    session_text(
                        "transcript",
                        "omitted_messages",
                        omitted_messages=omitted_messages,
                    )
                )
            ]
            if omitted_messages > 0
            else []
        )
        + kept
    )


def session_tokens(tx: Transcript) -> int:
    return sum(map(_message_tokens, tx["messages"]))


def prepared_tokens(
    tx: Transcript,
    model: str | None = None,
    todos: list[dict[str, str]] | None = None,
) -> int:
    return sum(map(_message_tokens, prepared_messages(tx, model=model, todos=todos)))


_tokenizer: tiktoken.Encoding | None = None


def wait(label: str) -> Wait:
    return {"label": label, "active": False}


def start_wait(item: Wait) -> None:
    item["active"] = True
    rt._note(item["label"], tag="wait")


def stop_wait(item: Wait) -> None:
    item["active"] = False


def update_wait(item: Wait, label: str) -> None:
    item["label"] = label
    if item["active"]:
        rt._note(item["label"], tag="wait")


def log_wait(item: Wait, message: str) -> None:
    _ = item
    rt._note(message, tag="wait")


def _normalized_vote_text(text: str) -> str:
    return " ".join(text.split()).strip().lower()



def run_turn(
    client,
    transcript: Transcript,
    state: AgentState,
    model_spec,
    tool_definitions,
):
    _, model = rt.split_model_spec(model_spec)
    step = 0
    while True:
        note_progress(state)
        prepared = prepared_messages(transcript, model=model, todos=state["todos"])
        rt._debug_log(
            "request",
            model=model_spec,
            step=step,
            messages=[rt._msg_to_dict(message) for message in prepared],
            tool_count=len(tool_definitions),
        )
        size_str = rt.format_tokens(sum(map(_message_tokens, prepared)))
        spinner = wait(f"Waiting for {model_spec} | {size_str}")
        start_wait(spinner)

        def on_retry(attempt, max_attempts, error_ctx=None):
            excerpt = ""
            if error_ctx:
                excerpt = " | ".join(
                    line.strip()
                    for line in error_ctx.strip().splitlines()[:3]
                    if line.strip()
                )
            log_wait(
                spinner,
                f"retry {attempt}/{max_attempts}{': ' + excerpt if excerpt else ''}",
            )
            update_wait(
                spinner,
                f"Retrying {model_spec} (attempt {attempt}/{max_attempts}) | {size_str}",
            )

        try:
            if remaining_unattended_seconds(state) <= 0:
                raise TimeoutError(
                    "reached unattended timeout "
                    f"({rt._format_duration(state['unattended_limit_seconds'])}) without a final response"
                )
            message = client["chat_completion"](
                model=model,
                messages=prepared,
                tools=tool_definitions,
                tool_choice="auto",
                on_retry=on_retry,
            )
        finally:
            stop_wait(spinner)
        rt._debug_log(
            "response",
            model=model_spec,
            step=step,
            assistant=rt._msg_to_dict(message),
        )
        calls = list(message["tool_calls"])
        if calls:
            add_assistant(transcript, message)
            results = [
                {
                    "call_id": call["id"],
                    "name": call["name"],
                    "result": invoke_tool(
                        state["tool_registry"], state, call["name"], call["arguments"]
                    ),
                }
                for call in calls
            ]
            rt._debug_log(
                "tool_results",
                model=model_spec,
                step=step,
                results=[
                    {
                        "call_id": result["call_id"],
                        "name": result["name"],
                        "ok": result["result"]["ok"],
                    }
                    for result in results
                ],
            )
            add_tool_results(transcript, results)
            step += 1
            continue
        rt._print(value=message["content"])
        return 0, message["content"]


def run_agent(
    prompt,
    model,
    root,
    system_prompt,
    unattended_limit_seconds,
    interactive,
    yolo: bool = False,
    transcript: Transcript | None = None,
):
    tool_registry = active_tool_registry(interactive)
    unattended_limit_seconds = _positive_int(
        unattended_limit_seconds, "unattended_limit_seconds"
    )
    state = new_agent_state(
        root=root,
        tool_registry=tool_registry,
        unattended_limit_seconds=unattended_limit_seconds,
        interactive=interactive,
        yolo=yolo,
    )
    if transcript is None:
        transcript = transcript_with_system_prompt(system_prompt)
    else:
        set_system_prompt(transcript, system_prompt)
    add_user(transcript, prompt)

    def runner(client):
        return run_turn(
            client,
            transcript,
            state,
            model,
            tool_specs(tool_registry),
        )

    try:
        return runner(rt.get_client(model))
    except (AuthenticationError, PermissionDeniedError) as exc:
        if not rt.ensure_api_env(root):
            return rt.fail(
                f"API {'authentication' if isinstance(exc, AuthenticationError) else 'permission'} error: {exc}"
            ), ""
        rt._warn("Credentials expired. Refreshing.")
        try:
            return runner(rt.get_client(model))
        except (AuthenticationError, PermissionDeniedError) as exc:
            return rt.fail(
                f"API {'authentication' if isinstance(exc, AuthenticationError) else 'permission'} error: {exc}"
            ), ""
        except Exception as exc:
            return rt.fail(str(exc)), ""
    except RateLimitError as exc:
        return rt.fail(f"API rate limit: {exc}"), ""
    except BadRequestError as exc:
        return rt.fail(f"API bad request: {exc}"), ""
    except Exception as exc:
        return rt.fail(str(exc)), ""


__all__ = [
    "AgentState",
    "Transcript",
    "agent_state",
    "new_agent_state",
    "remaining_unattended_seconds",
    "note_progress",
    "transcript",
    "transcript_with_system_prompt",
    "set_system_prompt",
    "clear_transcript",
    "checkpoint",
    "rollback",
    "undo_last_turn",
    "add_user",
    "add_assistant",
    "add_tool_results",
    "prepared_messages",
    "session_tokens",
    "prepared_tokens",
    "_message_tokens",
    "_pack_messages_with_toons",
    "_packed_history_note",
    "run_agent",
    "run_turn",
]

File: internal/oy/agent/agent.go
package agent

import (
	"fmt"
	"sort"
	"strings"
	"time"

	"github.com/wagov-dtt/oy-cli/internal/oy/providers"
	"github.com/wagov-dtt/oy-cli/internal/oy/runtime"
	"github.com/wagov-dtt/oy-cli/internal/oy/tools"
)

type State struct {
	Root                    string
	ToolRegistry            map[string]tools.RegisteredTool
	UnattendedLimitSeconds  int
	UnattendedDeadline      time.Time
	Interactive             bool
	ApproveAllMutatingTools bool
	Yolo                    bool
	Todos                   []map[string]string
}

type Transcript struct {
	Messages         []providers.ChatMessage `json:"messages"`
	MaxContextTokens int                     `json:"max_context_tokens"`
	MaxMessageTokens int                     `json:"max_message_tokens"`
}

type CompletionRunner interface {
	ChatCompletion(model string, messages []providers.ChatMessage, specs []map[string]any, toolChoice string) (providers.ChatMessage, error)
}

var PrintFunc = func(value string) { fmt.Println(value) }
var NoteFunc = func(string) {}

func AgentState(root string, toolRegistry map[string]tools.RegisteredTool, unattendedLimitSeconds int, unattendedDeadline time.Time, interactive, approveAllMutatingTools, yolo bool, todos []map[string]string) State {
	copiedTodos := make([]map[string]string, 0, len(todos))
	for _, item := range todos {
		copied := map[string]string{}
		for key, value := range item {
			copied[key] = value
		}
		copiedTodos = append(copiedTodos, copied)
	}
	return State{
		Root:                    root,
		ToolRegistry:            toolRegistry,
		UnattendedLimitSeconds:  unattendedLimitSeconds,
		UnattendedDeadline:      unattendedDeadline,
		Interactive:             interactive,
		ApproveAllMutatingTools: approveAllMutatingTools,
		Yolo:                    yolo,
		Todos:                   copiedTodos,
	}
}

func NewAgentState(root string, toolRegistry map[string]tools.RegisteredTool, unattendedLimitSeconds int, interactive, yolo bool) State {
	return AgentState(
		root,
		toolRegistry,
		unattendedLimitSeconds,
		time.Now().Add(time.Duration(unattendedLimitSeconds)*time.Second),
		interactive,
		yolo,
		yolo,
		nil,
	)
}

func RemainingUnattendedSeconds(state State) float64 {
	return time.Until(state.UnattendedDeadline).Seconds()
}

func NoteProgress(state State) error {
	if RemainingUnattendedSeconds(state) <= 0 {
		return fmt.Errorf(
			"reached unattended timeout (%s) without a final response",
			formatDuration(state.UnattendedLimitSeconds),
		)
	}
	return nil
}

func TranscriptState(messages []providers.ChatMessage, maxContextTokens, maxMessageTokens int) Transcript {
	copied := append([]providers.ChatMessage(nil), messages...)
	return Transcript{Messages: copied, MaxContextTokens: maxContextTokens, MaxMessageTokens: maxMessageTokens}
}

func TranscriptWithSystemPrompt(systemPrompt string) Transcript {
	tx := TranscriptState(nil, runtime.MaxContextTokens(), runtime.DefaultBudgets().MessageTokens)
	SetSystemPrompt(&tx, systemPrompt)
	return tx
}

func SetSystemPrompt(tx *Transcript, systemPrompt string) {
	if len(tx.Messages) > 0 && tx.Messages[0].Role == "system" {
		tx.Messages[0] = providers.SystemMessage(systemPrompt)
		return
	}
	tx.Messages = append([]providers.ChatMessage{providers.SystemMessage(systemPrompt)}, tx.Messages...)
}

func ClearTranscript(tx *Transcript, systemPrompt string) {
	tx.Messages = nil
	SetSystemPrompt(tx, systemPrompt)
}

func Checkpoint(tx Transcript) int {
	return len(tx.Messages)
}

func Rollback(tx *Transcript, point int) {
	if point < 0 {
		point = 0
	}
	if point > len(tx.Messages) {
		point = len(tx.Messages)
	}
	tx.Messages = tx.Messages[:point]
}

func UndoLastTurn(tx *Transcript) bool {
	for index := len(tx.Messages) - 1; index > 0; index-- {
		if tx.Messages[index].Role == "user" {
			tx.Messages = tx.Messages[:index]
			return true
		}
	}
	return false
}

func AddUser(tx *Transcript, prompt string) {
	tx.Messages = append(tx.Messages, providers.UserMessage(prompt))
}

func AddAssistant(tx *Transcript, message providers.ChatMessage) {
	tx.Messages = append(tx.Messages, message)
}

func AddToolResults(tx *Transcript, results []map[string]any) {
	for _, result := range results {
		callID, _ := result["call_id"].(string)
		name, _ := result["name"].(string)
		toolResult, _ := result["result"].(providers.ToolResult)
		tx.Messages = append(tx.Messages, providers.ToolMessage(callID, name, toolResult))
	}
}

func PreparedMessages(tx Transcript, todos []map[string]string) []providers.ChatMessage {
	messages := append([]providers.ChatMessage(nil), tx.Messages...)
	systemMessages := []providers.ChatMessage{}
	for _, message := range messages {
		if message.Role == "system" {
			systemMessages = append(systemMessages, message)
		}
	}
	if len(todos) > 0 {
		todoText, _ := runtime.SessionText(map[string]string{"todos": formatTodos(todos)}, "transcript", "todo_system")
		systemMessages = append(systemMessages, providers.SystemMessage(strings.TrimSpace(todoText)))
	}
	other := []providers.ChatMessage{}
	for _, message := range messages {
		if message.Role != "system" {
			other = append(other, message)
		}
	}
	budget := tx.MaxContextTokens - len(systemMessages)
	if budget <= 0 {
		return systemMessages
	}
	kept := []providers.ChatMessage{}
	if len(other) > budget {
		omitted := len(other) - budget
		omittedText, _ := runtime.SessionText(map[string]string{"omitted_messages": fmt.Sprintf("%d", omitted)}, "transcript", "omitted_messages")
		kept = append(kept, providers.UserMessage(omittedText))
		other = other[len(other)-budget:]
	}
	kept = append(kept, other...)
	return append(systemMessages, kept...)
}

func SessionTokens(tx Transcript) int {
	return len(tx.Messages)
}

func PreparedTokens(tx Transcript, todos []map[string]string) int {
	return len(PreparedMessages(tx, todos))
}

func ChooseSelfConsistentMessage(messages []providers.ChatMessage) (providers.ChatMessage, int, int, error) {
	if len(messages) == 0 {
		return providers.ChatMessage{}, 0, 0, fmt.Errorf("messages must not be empty")
	}
	if len(messages) == 1 {
		return messages[0], 0, 1, nil
	}
	bestIndex := 0
	bestVotes := -1
	bestSupport := -1
	for index, candidate := range messages {
		votes := 0
		support := 0
		for _, other := range messages {
			score := messageMatchScore(candidate, other)
			support += score
			if score >= 85 {
				votes++
			}
		}
		if votes > bestVotes || (votes == bestVotes && (support > bestSupport || (support == bestSupport && index < bestIndex))) {
			bestIndex = index
			bestVotes = votes
			bestSupport = support
		}
	}
	return messages[bestIndex], bestIndex, bestVotes, nil
}

func RunTurn(client CompletionRunner, transcript *Transcript, state *State, modelSpec string, toolDefinitions []map[string]any, bestOf int) (int, string, error) {
	_, model := providers.SplitModelSpec(modelSpec)
	if bestOf <= 0 {
		return 1, "", fmt.Errorf("best_of must be a positive integer")
	}
	step := 0
	for {
		if err := NoteProgress(*state); err != nil {
			return 1, "", err
		}
		prepared := PreparedMessages(*transcript, state.Todos)
		runtime.DebugLog("request", map[string]any{
			"model":      modelSpec,
			"step":       step,
			"messages":   debugMessages(prepared),
			"tool_count": len(toolDefinitions),
		})
		responses := make([]providers.ChatMessage, 0, bestOf)
		for i := 0; i < bestOf; i++ {
			message, err := client.ChatCompletion(model, prepared, toolDefinitions, "auto")
			if err != nil {
				return 1, "", err
			}
			responses = append(responses, message)
		}
		message, chosenIndex, voteCount, err := ChooseSelfConsistentMessage(responses)
		if err != nil {
			return 1, "", err
		}
		runtime.DebugLog("response", map[string]any{
			"model":     modelSpec,
			"step":      step,
			"assistant": debugMessage(message),
		})
		if bestOf > 1 && voteCount < bestOf {
			NoteFunc(fmt.Sprintf("self-consistency: sample %d won %d/%d", chosenIndex+1, voteCount, bestOf))
		}
		if len(message.ToolCalls) > 0 {
			NoteFunc(fmt.Sprintf("turn %d: %s", step+1, tools.CountText(len(message.ToolCalls), "tool call", "")))
			AddAssistant(transcript, message)
			results := make([]map[string]any, 0, len(message.ToolCalls))
			for _, call := range message.ToolCalls {
				NoteFunc("tool " + formatToolCallPreview(call))
				toolState := &tools.State{
					Root:                    state.Root,
					Interactive:             state.Interactive,
					Yolo:                    state.Yolo,
					ApproveAllMutatingTools: state.ApproveAllMutatingTools,
					Todos:                   cloneTodos(state.Todos),
				}
				result := tools.InvokeTool(state.ToolRegistry, toolState, call.Name, call.Arguments)
				NoteFunc(fmt.Sprintf("tool %s: %s", call.Name, formatToolResultPreview(result)))
				state.ApproveAllMutatingTools = toolState.ApproveAllMutatingTools
				state.Todos = cloneTodos(toolState.Todos)
				results = append(results, map[string]any{"call_id": call.ID, "name": call.Name, "result": result})
			}
			runtime.DebugLog("tool_results", map[string]any{
				"model":   modelSpec,
				"step":    step,
				"results": debugToolResults(results),
			})
			AddToolResults(transcript, results)
			step++
			continue
		}
		PrintFunc(message.Content)
		return 0, message.Content, nil
	}
}

func RunAgent(client CompletionRunner, prompt, model, root, systemPrompt string, unattendedLimitSeconds int, interactive bool, yolo bool, transcript *Transcript, bestOf int) (int, string, error) {
	toolRegistry := tools.ActiveToolRegistry(interactive)
	if unattendedLimitSeconds <= 0 {
		return 1, "", fmt.Errorf("unattended_limit_seconds must be a positive integer")
	}
	if bestOf <= 0 {
		return 1, "", fmt.Errorf("best_of must be a positive integer")
	}
	state := NewAgentState(root, toolRegistry, unattendedLimitSeconds, interactive, yolo)
	if transcript == nil {
		tx := TranscriptWithSystemPrompt(systemPrompt)
		transcript = &tx
	} else {
		SetSystemPrompt(transcript, systemPrompt)
	}
	AddUser(transcript, prompt)
	return RunTurn(client, transcript, &state, model, tools.ToolSpecs(toolRegistry), bestOf)
}

func cloneTodos(todos []map[string]string) []map[string]string {
	out := make([]map[string]string, 0, len(todos))
	for _, item := range todos {
		copied := map[string]string{}
		for key, value := range item {
			copied[key] = value
		}
		out = append(out, copied)
	}
	return out
}

func debugMessage(message providers.ChatMessage) map[string]any {
	payload := map[string]any{"role": message.Role}
	if message.Content != "" {
		payload["content"] = message.Content
	}
	if len(message.ToolCalls) > 0 {
		calls := make([]map[string]any, 0, len(message.ToolCalls))
		for _, call := range message.ToolCalls {
			entry := map[string]any{"id": call.ID, "name": call.Name}
			if len(call.Arguments) > 0 {
				entry["arguments"] = call.Arguments
			}
			calls = append(calls, entry)
		}
		payload["tool_calls"] = calls
	}
	if len(message.ThoughtSignatures) > 0 {
		payload["thought_signatures"] = message.ThoughtSignatures
	}
	if message.ToolCallID != "" {
		payload["tool_call_id"] = message.ToolCallID
	}
	if message.Name != "" {
		payload["name"] = message.Name
	}
	return payload
}

func debugMessages(messages []providers.ChatMessage) []map[string]any {
	out := make([]map[string]any, 0, len(messages))
	for _, message := range messages {
		out = append(out, debugMessage(message))
	}
	return out
}

func debugToolResults(results []map[string]any) []map[string]any {
	out := make([]map[string]any, 0, len(results))
	for _, result := range results {
		entry := map[string]any{"call_id": result["call_id"], "name": result["name"]}
		if toolResult, ok := result["result"].(providers.ToolResult); ok {
			entry["ok"] = toolResult.OK
		}
		out = append(out, entry)
	}
	return out
}

func formatToolCallPreview(call providers.ToolCall) string {
	if len(call.Arguments) == 0 {
		return call.Name
	}
	keys := make([]string, 0, len(call.Arguments))
	for key := range call.Arguments {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	details := make([]string, 0, len(keys))
	for _, key := range keys {
		details = append(details, fmt.Sprintf("%s=%s", key, runtime.Preview(call.Arguments[key], 80)))
	}
	return fmt.Sprintf("%s(%s)", call.Name, strings.Join(details, ", "))
}

func formatToolResultPreview(result providers.ToolResult) string {
	status := "error"
	if result.OK {
		status = "ok"
	}
	preview := strings.TrimSpace(runtime.Preview(result.Content, 120))
	if preview == "" {
		return status
	}
	return fmt.Sprintf("%s %s", status, preview)
}

func normalizedVoteText(text string) string {
	clean := strings.Map(func(r rune) rune {
		if (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') || (r >= '0' && r <= '9') {
			return r
		}
		return ' '
	}, text)
	return strings.ToLower(strings.Join(strings.Fields(clean), " "))
}

func textMatchScore(left, right string) int {
	if normalizedVoteText(left) == normalizedVoteText(right) {
		return 100
	}
	leftWords := strings.Fields(normalizedVoteText(left))
	rightWords := strings.Fields(normalizedVoteText(right))
	if len(leftWords) == 0 || len(rightWords) == 0 {
		return 0
	}
	matches := 0
	seen := map[string]struct{}{}
	for _, word := range leftWords {
		seen[word] = struct{}{}
	}
	for _, word := range rightWords {
		if _, ok := seen[word]; ok {
			matches++
		}
	}
	maxWords := len(leftWords)
	if len(rightWords) > maxWords {
		maxWords = len(rightWords)
	}
	return (matches * 100) / maxWords
}

func toolPlanMatchScore(left, right []providers.ToolCall) int {
	if len(left) == 0 || len(right) == 0 {
		if len(left) == len(right) {
			return 100
		}
		return 0
	}
	if len(left) != len(right) {
		return 0
	}
	score := 0
	for i := range left {
		if left[i].Name != right[i].Name {
			return 0
		}
		if providers.SerializeJSON(left[i].Arguments) == providers.SerializeJSON(right[i].Arguments) {
			score += 100
			continue
		}
		score += textMatchScore(providers.SerializeJSON(left[i].Arguments), providers.SerializeJSON(right[i].Arguments))
	}
	return score / len(left)
}

func messageMatchScore(left, right providers.ChatMessage) int {
	leftCalls := left.ToolCalls
	rightCalls := right.ToolCalls
	if (len(leftCalls) == 0) != (len(rightCalls) == 0) {
		return 0
	}
	if len(leftCalls) > 0 {
		plan := toolPlanMatchScore(leftCalls, rightCalls)
		if left.Content == "" && right.Content == "" {
			return plan
		}
		return (85*plan + 15*textMatchScore(left.Content, right.Content)) / 100
	}
	return textMatchScore(left.Content, right.Content)
}

func formatDuration(seconds int) string {
	if seconds%3600 == 0 {
		return fmt.Sprintf("%dh", seconds/3600)
	}
	if seconds%60 == 0 {
		return fmt.Sprintf("%dm", seconds/60)
	}
	return fmt.Sprintf("%ds", seconds)
}

func formatTodos(todos []map[string]string) string {
	lines := make([]string, 0, len(todos))
	for _, item := range todos {
		status := item["status"]
		icon := map[string]string{"pending": "[ ]", "in_progress": "[~]", "done": "[x]"}[status]
		if icon == "" {
			icon = "[ ]"
		}
		line := fmt.Sprintf("%s %s: %s", icon, strings.TrimSpace(item["id"]), strings.TrimSpace(item["task"]))
		lines = append(lines, line)
	}
	return strings.Join(lines, "\n")
}
